### DISCUSSION TITLE
How to grab all IPs associated to a device_id in custom script

### DISCUSSION THREAD

**Discussion by WhaleJ84** [netbox-community/netbox#12860] [Q&A] [answered]
Hey all, currently writing a script that will automatically generate OpenBSD interface configs for devices. I've got a script that allows you to select a device and then an interface to generate for. However once the script is ran, I am not sure how to grab all IPs associated to the selected device_id for use within the config.

Here's my current WIP code:
```python
class InterfaceDataScript(Script):
    class Meta:
        name = "Interface Data"
        description = "Test"
        commit_default = False

    device = ObjectVar(
        Device
    )
    interface = ObjectVar(
        Interface,
        query_params={
            'device_id': '$device'
        }
    )


    def run(self, data, commit):
        device = data['device']
        ipaddress = IPAddress.objects.filter(assigned_object_id=device.id)
        interface = data['interface']
        label = interface.label
        file = f"/etc/hostname.{label}"
        enabled = ""
        if interface.enabled:
            enabled = "up"

        return f"""DEBUG:
{vars(interface)}
{vars(device)}
{vars(ipaddress)}

# {label}
{enabled}
"""
```

I will add in logic to detect whether IP addresses are even assigned to the device later, but for now I just want all IP related data for the device saved into the `ipaddress` variable.
Is someone able to help me achieve this? I'm poking around the documentation and not seeing the answer.

**Comment by WhaleJ84** [comment]
I guess another way of phrasing it is 'How to I get the same data as hitting the API endpoint `api/ipam/ip-addresses/?device_id=$DEVICE_ID` into my script above?'

**Reply by candlerb** [reply]
The best way to experiment with the Django ORM is via [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/):

```
/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
```

IPAddress objects are linked to Interfaces (or VMInterfaces or FHRP Groups), and in turn Interfaces are linked to Devices.  The IPAddress isn't linked directly to the device, which is why your query failed.

> I guess another way of phrasing it is 'How to I get the same data as hitting the API endpoint `api/ipam/ip-addresses/?device_id=$DEVICE_ID` into my script above?'

You can find the source for that REST API view and filter here:
* https://github.com/netbox-community/netbox/blob/v3.5.3/netbox/ipam/views.py#L719-L723  -->
* https://github.com/netbox-community/netbox/blob/v3.5.3/netbox/ipam/filtersets.py#L545-L549  -->
* https://github.com/netbox-community/netbox/blob/v3.5.3/netbox/ipam/filtersets.py#L670-L679

Look at the last link in particular. It finds all devices with `device_id__in=[ids]` (since the REST API lets you repeat the device_id parameter), then finds all interfaces on those devices, then finds all IP addresses on those interfaces.

A simplified version (where you're not worried about virtual chassis) might be this:

```
>>> device = Device.objects.get(name="nuc1")
>>> interfaces = device.interfaces.all()   # OR: Interface.objects.filter(device=device)
>>> addrs = IPAddress.objects.filter(interface__in=interfaces)
# OR:
>>> addrs = []
>>> for i in interfaces: addrs.extend(i.ip_addresses.all())
```

This relies on the `interface` filter on IPAddress. If you want to use a raw assigned_object_id as in the code you showed, it would be something like this:

```
>>> addrs = IPAddress.objects.filter(assigned_object_id__in=[i.id for i in interfaces],
... assigned_object_type_id=ContentType.objects.get(app_label='dcim',model='interface').id)
```

(I don't know if there's an easier way to get from the "Interface" class to its django content_type_id)

**Reply by candlerb** [reply, marked as answer]
> (I don't know if there's an easier way to get from the "Interface" class to its django content_type_id)

Answering my own question:

```
>>> ContentType.objects.get_for_model(Interface)
<ContentType: dcim | interface>
>>> ContentType.objects.get_for_model(Interface).id
17
```

You don't even need the id:

```
>>> IPAddress.objects.filter(assigned_object_id__in=[i.id for i in interfaces],
    assigned_object_type=ContentType.objects.get_for_model(Interface))
```

Unfortunately, you can't do the following:

```
>>> IPAddress.objects.filter(assigned_object__in=interfaces)
...
django.core.exceptions.FieldError: Field 'assigned_object' does not generate an automatic reverse relation and
therefore cannot be used for reverse querying. If it is a GenericForeignKey, consider adding a GenericRelation.
```

**Reply by WhaleJ84** [reply]
Thanks for your in-depth reply! I've tried it on my end and although it does return results, it returns RestrictedQuerySet objects that doesn't have any of the attributes I'm expecting (e.g. `id`, `primary_ip4`, `FQDN`, etc.). Am I doing it wrong or are there more steps required from there?

**Reply by candlerb** [reply]
Depends what you want to do. If you want to iterate over it, it should be fine. If you want to turn it into a list of concrete objects, wrap with `list(...)`

**Reply by candlerb** [reply]
`primary_ip4` is an attribute of Device, not IPAddress. Address has `dns_name`, not `FQDN`. Inside nbshell `help(IPAddress)` or `help(Device)` will give you a full list of attributes.

```
>>> device = Device.objects.get(name="nuc1")
>>> interfaces = device.interfaces.all()
>>> addrs = IPAddress.objects.filter(interface__in=interfaces)
>>> addrs[0]
<IPAddress: 10.12.12.1/24>
>>> addrs[0].dns_name
''
>>> device.primary_ip4
<IPAddress: 10.12.255.11/24>
```

**Reply by WhaleJ84** [reply]
Perfect, thank you!
